### PR TITLE
Pass sampleUnitId through to case service

### PR DESCRIPTION
**Pull request by madeye-matt**
# Motivation and Context
Originally this was going to be a simple change to include the sample UUID (newly passed in the message from the sample service) in the message that gets passed in the case service.  

However, we strongly felt that an integration test should be created to test these changes (i.e. one where the sample message is passed in by the test which then asserts that a message has been sent to the case service).

In order for the message to be sent to the case service, the sample must be fully validated.  Also the collection exercise must be in the VALIDATED state, which requires it to be executed first.  For this to work end to end, the collection exercise service needed to be changed to fully support social surveys.  In the end, this required surprisingly little code but much investigation (details in the next section).

# What has changed
- sample unit id (UUID) now received in message from sample service and persisted in the collection exercise service internal representation
- when a collection exercise is created for a social survey, the correct action plans, case type defaults etc are now created
- for social surveys, no attempt is made to get a party id for the incoming samples 
- samples for social surveys are validated by mandating a collection instrument but insisting that party is absent (null)
- a new integration test ([ensureSampleUnitIdIsPropagatedHere](https://github.com/ONSdigital/rm-collection-exercise-service/blob/5e2c233d5ee7102aa56fcafa18ef9ba4be2d547b/src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java#L207)) has been added to test these changes end to end 
- mocking using wiremock has been added to allow the collection exercise service to get details of collection instrument (party service response also added to get tests to pass initially but not currently used - this has been left in because we will need to make a similar IT for business surveys and it will be needed for that)
- test data and unit tests have been updated to account for the additional sample unit id

# How to test?
- check out the code
- run 'mvn clean install'
- ensure all tests pass

# TODO 
- implement similar integration test for business surveys

# Links
[New integration test](https://github.com/ONSdigital/rm-collection-exercise-service/blob/5e2c233d5ee7102aa56fcafa18ef9ba4be2d547b/src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java#L207)
[WireMock](http://wiremock.org/)
[Trello card](https://trello.com/c/ZFF7x8Za/102-dev-task-3b-int-create-action-service-to-retrieve-new-sample-attributes-13)

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java [hunk 1]
```diff
@@ -424,27 +454,20 @@ private CollectionExerciseEvent getCollectionExerciseTransitionState(CollectionE
...
 
-    Predicate<ExerciseSampleUnit> stateTest = su -> su.getPartyId() instanceof UUID
-        && su.getCollectionInstrumentId() instanceof UUID;
-    boolean stateValidated = false;
-    if (sampleUnits.isEmpty()) {
-      stateValidated = false;
-    } else {
-      stateValidated = sampleUnits.stream().allMatch(stateTest);
-    }
+    boolean stateValidated = sampleUnits.size() > 0 && sampleUnits.stream().allMatch(ValidateSampleUnits::isSampleUnitValid);
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java]: Much better

File: src/main/resources/database/changes/release-10.49.2/add-id-to-sampleunit.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+ALTER TABLE collectionexercise.sampleunit ADD COLUMN id UUID CONSTRAINT su_id_uuid UNIQUE;
```
Comment (benjefferies) [inline src/main/resources/database/changes/release-10.49.2/add-id-to-sampleunit.sql]: Is there no chance the same sample would be linked to multiple CE?

File: src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java [hunk 3]
```diff
@@ -1,162 +1,344 @@
...
-
-        assertEquals(1, links.size());
-
-        SampleLinkDTO link = links.get(0);
-
-        assertEquals(sampleSummaryId, UUID.fromString(link.getSampleSummaryId()));
-        assertEquals("ACTIVE", link.getState());
-    }
-}
+  // TODO pull these from config
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java]: Do we need this comment?

File: src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java [hunk 4]
```diff
@@ -1,162 +1,344 @@
...
+  @Autowired
+  private ObjectMapper mapper;
+
+  private ObjectMapper xmlMapper = new XmlMapper();
+
+  @Autowired
+  private AppConfig appConfig;
+
+  @ClassRule
+  public static final SpringClassRule SPRING_CLASS_RULE = new SpringClassRule();
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java]: Is this needed?

File: src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java [hunk 5]
```diff
@@ -1,162 +1,344 @@
...
+
+  @Rule
+  public final SpringMethodRule springMethodRule = new SpringMethodRule();
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java]: Is this needed?

File: src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java [hunk 6]
```diff
@@ -1,162 +1,344 @@
...
+                                                  "collection-outbound-exchange", "Case.CaseDelivery.binding");
+
+    String xml = sampleUnitToXmlString(sampleUnit);
+
+    log.debug("xml = " + xml);
+
+    sender.sendMessage("sample-outbound-exchange", "Sample.SampleDelivery.binding",
+                       xml);
+
+    // The service stubs will exit once the call below times out preventing further debugging of this test in other
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java]: Not sure if this comment is necessary if we're using queue.take() in the test normally?

**Comment by emilioward** [comment]
[Dev task 3B: INT: Create action service to retrieve new sample attributes (13)](https://trello.com/c/ZFF7x8Za/102-dev-task-3b-int-create-action-service-to-retrieve-new-sample-attributes-13)

File: src/main/resources/database/changes/release-10.49.2/add-id-to-sampleunit.sql (hunk 2 above)
Comment (madeye-matt) [reply src/main/resources/database/changes/release-10.49.2/add-id-to-sampleunit.sql]: Not sure how that would affect this - all we are doing is adding a UUID field to the sample unit that gets passed through from sample service

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/SampleServiceImpl.java [hunk 7]
```diff
@@ -159,10 +160,12 @@ public ExerciseSampleUnit acceptSampleUnit(SampleUnit sampleUnit) throws CTPExce
                 exerciseSampleUnit = new ExerciseSampleUnit();
                 exerciseSampleUnit.setSampleUnitGroup(sampleUnitGroup);
                 exerciseSampleUnit.setSampleUnitRef(sampleUnit.getSampleUnitRef());
+                exerciseSampleUnit.setSampleUnitId(UUID.fromString(sampleUnit.getId()));
                 exerciseSampleUnit.setSampleUnitType(SampleUnitDTO.SampleUnitType.valueOf(sampleUnit.getSampleUnitType()));
 
                 sampleUnitRepo.saveAndFlush(exerciseSampleUnit);
 
+                // null pointer exception thrown here
```
[the remark this answers is not in the document]
Comment (madeye-matt) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/SampleServiceImpl.java]: Was a debugging comment  - now removed

**Comment by benjefferies** [comment]
Looks like the integration tests aren't running in travis

File: src/main/resources/database/changes/release-10.49.2/add-id-to-sampleunit.sql (hunk 2 above)
Comment (benjefferies) [reply src/main/resources/database/changes/release-10.49.2/add-id-to-sampleunit.sql]: Assuming we aren't going to link the exact same sample this should be ok

**Comment by benjefferies** [comment]
I'm getting errors when trying to run the acceptance tests locally with this branch

```
java.lang.NullPointerException: null
        at uk.gov.ons.ctp.response.collection.exercise.service.impl.CollectionExerciseServiceImpl.createActionPlans(CollectionExerciseServiceImpl.java:287)
        at uk.gov.ons.ctp.response.collection.exercise.service.impl.CollectionExerciseServiceImpl.createCollectionExercise(CollectionExerciseServiceImpl.java:256)
``

**Comment by benjefferies** [comment]
@madeye-matt the acceptance tests still appear to hang even with this change

**Comment by tej99** [comment]
@benjefferies they stop hanging at the initial collection execution stage with https://github.com/ONSdigital/rm-case-service/pull/49#pullrequestreview-132315756
but they start hanging later in the acceptance tests now.

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/distribution/SampleUnitDistributor.java [hunk 8]
```diff
@@ -144,16 +142,18 @@ private void distributeSampleUnits(CollectionExercise exercise, ExerciseSampleUn
...
+        parent.setId(sampleUnit.getSampleUnitId().toString());
+        parent.setPartyId(Objects.toString(sampleUnit.getPartyId(), null));
         parent.setCollectionInstrumentId(sampleUnit.getCollectionInstrumentId().toString());
         actionPlanId = collectionExerciseRepo
             .getActiveActionPlanId(exercise.getExercisePK(), sampleUnit.getSampleUnitType().name(),
                 exercise.getSurveyId());
       } else {
         SampleUnitChild child = new SampleUnitChild();
         child.setSampleUnitRef(sampleUnit.getSampleUnitRef());
+        child.setId(sampleUnit.getSampleUnitId().toString());
```
Comment (tej99) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/distribution/SampleUnitDistributor.java]: Getting a NullPointerException around here when trying to execute collection exercises?

File: src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java [hunk 9]
```diff
@@ -205,9 +206,36 @@ public void shouldActivateSampleLink() throws Exception {
...
+    log.info("************ Starting ensureSampleUnitIdIsPropagatedHereSocial *************");
     createSurveyServiceSocialStub();
+    SampleUnitParent sampleUnit = ensureSampleUnitIdIsPropagatedHere("H");
+
+    assertNull("Party id must be null", sampleUnit.getPartyId());
+  }
+
+  @Test
+  public void ensureSampleUnitIdIsPropagatedHereBusiness() throws Exception {
+    log.info("************ Starting ensureSampleUnitIdIsPropagatedHereBusiness *************");
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java]: Did you commit these logs by accident?

File: src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java [hunk 10]
```diff
@@ -205,9 +206,36 @@ public void shouldActivateSampleLink() throws Exception {
...
+    createSurveyServiceBusinessStub();
+    createPartyServiceNoAssociationsStub();
+    SampleUnitParent sampleUnit = ensureSampleUnitIdIsPropagatedHere("B");
+
+    assertNotNull("Party id must be not null", sampleUnit.getPartyId());
+  }
+
+  @Test
+  public void ensureSampleUnitIdIsPropagatedHereBusinessWithExistingEnrolments() throws Exception {
+    log.info("************ Starting ensureSampleUnitIdIsPropagatedHereBusinessWithExistingEnrolments *************");
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java]: Did you commit these logs by accident?

**Pull request merged**